torch/_inductor/decomposition.py

```diff
@@ -332,15 +332,49 @@ def angle(x):
 
 @register_decomposition([aten.add])
 def add(x, y, *, alpha=None):
-    x_is_complex_tensor = torch.is_tensor(x) and x.is_complex()
-    y_is_complex_tensor = torch.is_tensor(y) and y.is_complex()
-    if not x_is_complex_tensor or not y_is_complex_tensor:
+    # Require both x and y to be tensors and at least one of them is complex typed.
+    if (not torch.is_tensor(x) or not torch.is_tensor(y)) or (
+        not x.is_complex() and not y.is_complex()
+    ):
         return NotImplemented
     z = y
     if alpha is not None:
         z = alpha * y
     complex_type = torch.promote_types(x.dtype, y.dtype)
-    return (x.view(x.real.dtype) + z.view(y.real.dtype)).view(complex_type)
+
+    # For complex typed `x`, `x.view(x.real.dtype)` doubles the last dimension and can cause problem
+    # when broadcasting the add.
+    def reshape_tensor_complex(tensor):
+        """Reshape tensor from [*initial_dims, last_dim] to *initial_dims, last_dim/2, 2]"""
+        # Get the current shape of the tensor
+        *initial_dims, last_dim = tensor.shape
+
+        # Check if the last dimension is even. We should never reach here since `x.view(x.real.dtype)`
+        # doubles the last dimension for complex numbers.
+        if last_dim % 2 != 0:
+            raise AssertionError(
+                "The size of the last dimension must be even to reshape it to [..., last_dim/2, 2]"
+            )
+
+        # Reshape the tensor
+        new_shape = (*initial_dims, last_dim // 2, 2)
+        reshaped_tensor = tensor.view(new_shape)
+        return reshaped_tensor
+
+    def reshape_tensor_real(tensor):
+        tensor_unsqueezed = tensor.unsqueeze(-1)
+        zeros = torch.zeros_like(tensor_unsqueezed)
+        return torch.cat((tensor_unsqueezed, zeros), dim=-1)
```

Comment:
Ah right. We have to do this. This is not going to fuse well, and we might even regress perf.
I can think of three ways of doing this:
- Doing `mask = (torch.arange(last_dim) % 2 == 0).view(last_dim // 2, 2)` and broadcasting the real tensor and choosing the sum or this tensor depending on this mask.
- A better way to do this at the moment would be via lowering with masked loads to avoid unnecessary compute
- Long term, we would represent complex numbers as two tensors (real and imaginary part) and all this would be just trivial to do

All these are a bit involved. Perhaps it's better to simply fallback to eager in this case and wait until someone tackles the general complex support problem.

Reply:
I see. To keep things simple, I'm now letting any add that involves real tensor fallback to eager.

Reply:
hmm ok, looks like if we let adding complex and real tensor fallback to eager, then dynamic codegen breaks. For example, the test below fails (my newly added test case).

  python test/inductor/test_torchinductor_codegen_dynamic_shapes.py -k DynamicShapesCodegenCpuTests.test_add_complex9_dynamic_shapes_cpu
  
  If I add back the decomposition, it would pass. 
  
  Is this expected? I think we still need a decomposition for adding complex number and real number?
  
  cc @zou3519

Reply:
Do these tests fail before your PR? (My guess is yes). If so, then I'd just skip them

Reply:
Yes. They do. I'll remove these tests then.

Reply:
Could you file some issues for them please?